### ISSUE TITLE
Remove Filters in incoming Request

### ISSUE THREAD

**Issue by rachael-ross** [json-api-dotnet/JsonApiDotNetCore#1182]
#### SUMMARY
Outside of implementing custom filters, I'm wondering if there is a way to **remove a filter expression** (and in this case add a different one).

#### DETAILS

I have an entity that has a StartDate and EndDate. However, I'd like to give the client app a way to query by Quarter and Year instead and read those values and transform the query expression to one that queries the StartDate and EndDate appropriately.

I've added Quarter and Year properties as attributes, but NotMapped: 

```
 public class PlanManagement : Identifiable<long>
    {
     
        [Attr]        
        public long? PlanId { get; set; }
    
        [HasOne]
        public Plan? Plan { get; set; }
      
        [Attr]
        public DateTime StartDate { get; set; }
       
        [Attr]
        public DateTime? EndDate { get; set; }
       
        [Attr, NotMapped]
        public short Quarter { get; set; }
       
       [Attr, NotMapped]
        public short Year { get; set; }

        . . .
}
```

Then, I created a PlanManagementExpressionRewriter:

```
 public class PlanManagementExpressionRewriter : QueryExpressionRewriter<object?>
 {
        public override QueryExpression? VisitComparison(ComparisonExpression expression, object? argument)
        { . . . }

        public override QueryExpression? VisitLogical(LogicalExpression expression, object? argument)
        { . . . }
}
```
and PlanManagementRepository:

```
public class PlanManagementRepository : EntityFrameworkRepository<PlanManagement, long>
    {
            private readonly PlanManagementExpressionRewriter _writer;

            public PlanManagementRepository(ITargetedFields targetedFields,
                                   IDbContextResolver contextResolver,
                                   IResourceGraph resourceGraph,
                                   IResourceFactory resourceFactory,
                                   IEnumerable<IQueryConstraintProvider> constraintProviders,
                                   IResourceDefinitionAccessor resourceDefinitionAccessor,
                                   ILoggerFactory loggerFactory,
                                   IConfiguration configuration,
                                   ICurrentUserAuthorizationService authService) 
            : base(targetedFields, contextResolver, resourceGraph, resourceFactory, constraintProviders, loggerFactory, resourceDefinitionAccessor)
         {
              _writer = new PlanManagementExpressionRewriter(resourceGraph);
         }

        protected override IQueryable<PlanManagement> ApplyQueryLayer(QueryLayer queryLayer)
        {
           if (queryLayer.Filter != null)
           {
               queryLayer.Filter = (FilterExpression?)_writer.Visit(queryLayer.Filter, null);
           }          
        }

        . . . 
}
```

1. Is this the best approach or can I do this in the ResourceDefinition:::OnApplyFilter()?
2. If this is the best approach, the issue I'm having is that the PlanManagementRepository::ApplyQueryLayer() doesn't fire if the PlanManagement records are being retrieved (and filtered) as a secondary resource (not the primary resource being requested).

#### VERSIONS USED
- JsonApiDotNetCore version: 5
- ASP.NET Core version: 6
- Entity Framework Core version: 6
- Database provider: Sql Server

**Comment by bkoelman** [comment]
This sounds like the right approach, but you'll need to call the rewriter from *all* repositories, not just the one for PlanManagement, so it handles non-primary expressions too. This can be done by adding a generic repository that derives from `EntityFrameworkCoreRepository<>` and registering it in IoC container using open generics (see below). In the rewriter, type-check for PlanManagement at the recursive virtual methods.

There's an example at https://github.com/json-api-dotnet/JsonApiDotNetCore/tree/master/test/JsonApiDotNetCoreTests/IntegrationTests/CompositeKeys that's somewhat similar. This sample uses a composite primary key (car region id + license plate) in the database and translates usage of JSON:API Id values in queries using a rewriter.

Registering the generic repository is done from startup like this:
```c#
builder.Services.AddScoped(typeof(IResourceReadRepository<,>), typeof(YourRepository<,>));
builder.Services.AddScoped(typeof(IResourceWriteRepository<,>), typeof(YourRepository<,>));
builder.Services.AddScoped(typeof(IResourceRepository<,>), typeof(YourRepository<,>));
```

The rewriter is quite an advanced topic, nice to hear you're diving in that deep and making use of these building blocks!

**Comment by rachael-ross** [comment]
Thank you, Bart! 

I've got all of that with the exception of moving the logic up to our generic repository. (I stated EntityFrameworkRepository above for simplicity). 

I mistakenly assumed the CarCompositeKeyAwareRepository was Car-specific due to the semantic naming of it, but now see it's a generic repo positioned to handle all types. It should've been a hint to me when I saw Car type was being tested for in IsCarId(). :upside_down_face: 

`CarCompositeKeyAwareRepository<TResource, TId>`
